### ISSUE TITLE
State inspector

### ISSUE THREAD

**Issue by rrousselGit** [rrousselGit/riverpod#1033]
Similar to provider, let's add a pannel on the Flutter devtool to inspect the state of a Riverpod app

**Comment by muzammil-git** [comment]
Hello Roussel, Thankyou for creating an amazing package.

Just wanted to know when will the full-blown inspector be coming for Riverpod. Any estimated dates/months ?

**Comment by rrousselGit** [comment]
It's coming. I'm just doing other things. Riverpod is being rewritten in big parts at the moment, to fix various bugs and implement certain features.

Making a devtool before finishing the refactor would just be inneficient.

**Comment by itinicore** [comment]
@rrousselGit any updates?

**Comment by binemmanuel** [comment]
@rrousselGit any update?

**Comment by pylvr** [comment]
any updates on this ?

**Comment by ishimwe-samuel** [comment]
@rrousselGit  any updates on this?

**Comment by BrianHub86** [comment]
?

**Comment by rrousselGit** [comment]
Don't spam.

I've made good progress. I'll share something soon.

**Comment by nateshmbhat** [comment]
any update on this ?

**Comment by rrousselGit** [comment]
Yes, I'm working on it. I'll show something by the end of the year.

To be fair, I already have a state inspector. I'm trying to build something a bit more useful than just that.
In particular, instead of just blindly showing the state (which you can obtain from the debugger), I want to show extra info such as "Why did this state change and who changed it"

**Comment by nateshmbhat** [comment]
this is great ! Looking forward to this 🤩 ❤️

**Comment by sweettirexkiller** [comment]
Any updates :D?

**Comment by rrousselGit** [comment]
Yep still on it. Got distracted by War Thunder for a bit x)

**Comment by rrousselGit** [comment]
I'm making good progress on this. 


The devtool allows you to "step back in time" to view previous states. I'm currently trying to figure out how to give the devtool superpowers.

Specifically I'd like to:
- Block all updates to a provider, essentially freezing it (therefore ignoring ref.watch & co).
- Show a diff between two different states
- Stub a provider (Kind of like overrides, but dynamically added from the devtool)

I'm also not super sure how to display events like "This provider got disposed" in a nice manner.
I essentially have a list of events over time. But showing the full list feels like too much info.

Anyway I'll share some screenshots soon.

**Comment by pylvr** [comment]
@rrousselGit is there a way to release a preview version. even if not stable it can be very usefull for many of us

**Comment by rrousselGit** [comment]
I'll consider it once it gets a little bit more stable.

**Comment by rrousselGit** [comment]
Quick demo
The circles in the bottom (state over time) will be improved.


https://github.com/user-attachments/assets/3daa9ed6-7a81-4414-9bf5-07f07225d608


I'm still figuring out how to display things I mentioned earlier (such as "When did this provider get disposed, why did it rebuild, ...).
My current though is to display that on over of the events in the bottom. Hovering one could show all the events that happened at that time.

**Comment by rrousselGit** [comment]
Oh and clicking on a provider jumps to its definition in the IDE

https://github.com/user-attachments/assets/536eaf6d-baac-49ed-98ed-76ed9f63590e

**Comment by rrousselGit** [comment]
Small readability improvement. I've separated modified providers from the others in the list.

And the "states over time" stepper at the bottom greys out all the steps unrelated to the selected provider. You can also hover it to see what's in a specific step

https://github.com/user-attachments/assets/44a14915-8df0-428d-b2f9-eeb52309f7e0

**Comment by rrousselGit** [comment]
I'm fixing a couple of important bugs and will ship the devtool alongside it.

**Comment by rrousselGit** [comment]
The bugfix is done. I'm improving the rendering of AsyncValue (which is unreadable at the moment), and I'll ship it. It should be out this week.

**Comment by rrousselGit** [comment]
Fwiw, the devtool is available when using `riverpod: 3.3.2-dev.2`

You may have to add `riverpod` to your `pubspec.yaml` explicitly. Although I am not sure if that's required yet :) 

<img width="1728" height="987" alt="Image" src="https://github.com/user-attachments/assets/56ca2b67-2387-4a47-b811-1bb30db9adec" />
